planning/motion_velocity_planner/autoware_motion_velocity_run_out_module/src/map_data.cpp

```diff
@@ -37,10 +37,33 @@ namespace autoware::motion_velocity_planner::run_out
 
 namespace
 {
+/// @brief return true if the given type string is contained in the vector of strings
 bool contains_type(const std::vector<std::string> & types, const std::string & type)
 {
   return std::find(types.begin(), types.end(), type) != types.end();
 }
+
+/// @brief return true if the given type and subtype strings match one of the string in the vector
+/// @details the strings in the vector are assumed to have format "type.subtype" or just "type"
+bool contains_type(
+  const std::vector<std::string> & types, const std::string & type, const std::string & subtype)
+{
+  if (subtype.empty()) {
+    return contains_type(types, type);
+  }
+  for (const auto & type_str : types) {
+    const auto dot_pos = type_str.find('.');
+    const auto main_type_str = type_str.substr(0, dot_pos);
+    if (main_type_str != type) {
+      continue;
+    }
+    const auto sub_type_str = type_str.substr(dot_pos + 1, type_str.size());
+    if (sub_type_str == subtype) {
+      return true;
+    }
+  }
+  return false;
+}
```

Comment:
```suggestion
bool contains_type(
  const std::vector<std::string> & types, const std::string & type, const std::string & subtype)
{
  if (subtype.empty()) {
    return std::find(types.begin(), types.end(), type) != types.end();
  }

  const std::string full_type = type + "." + subtype;

  for (const auto & type_str : types) {
    if (type_str == full_type) {
      return true;
    }

    const auto dot_pos = type_str.find('.');

    if (dot_pos == std::string::npos) {
      continue;
    }

    if (type_str.compare(0, dot_pos, type) != 0) {
      continue;
    }

    if (type_str.compare(dot_pos + 1, std::string::npos, subtype) == 0) {
      return true;
    }
  }
  return false;
}
```

The main purpose is to appeal that I've checked the code :sweat_smile: , but I'd like to propose some minor improvements. The current implementation is fine as is.
- By changing `substr()` to `compare()`, we can avoid string copying.
- By performing the exact match check first, we might improve the flow (though I'm not entirely confident about this).

Reply:
Thank you ! Your comment helped me simplify the function even more.